File: src/main/java/aoc2018/Day8.kt
package aoc2018

fun main() = Day8().run()

class Day8 : Day(8) {
    override fun part1(input: List<String>): Any {
        return Tree(input[0]).metaDataSum()
    }

    override fun part2(input: List<String>): Any {
        return Tree(input[0]).values()
    }


    private class Tree(input: String) {

        val numbers: List<Int>
        var idx = 0

        init {
            numbers = input.split(' ').map(String::toInt)
        }

        fun metaDataSum(): Int {
            var metaDataSum = 0
            val children = numbers[idx++]
            val metaDataCount = numbers[idx++]
            for (i in 0 until children) {
                metaDataSum += metaDataSum()
            }
            for (i in 0 until metaDataCount) {
                metaDataSum += numbers[idx++]
            }
            return metaDataSum
        }

        fun values(): Int {
            val children = numbers[idx++]
            val metaDataCount = numbers[idx++]
            val childValues = mutableMapOf<Int, Int>()
            for (i in 1..children) {
                childValues[i] = values()
            }
            var values = 0
            for (i in 1..metaDataCount) {
                val metaData = numbers[idx++]
                values += if (children == 0) {
                    metaData
                } else {
                    childValues[metaData] ?: 0
                }
            }
            return values
        }
    }
}

File: src/main/java/aoc2022/Day8.scala
package aoc2022

import scala.math.min

object Day8 extends Day(8) {

  def main(args: Array[String]): Unit = run()

  val trees: Seq[Seq[Int]] = input.map(_.split("").map(_.toInt))

  def treesAround(x: Int, y: Int): List[Seq[Int]] = {
    List(trees.take(y).map(_ (x)).reverse,
      trees.drop(y + 1).map(_ (x)),
      trees(y).take(x).reverse,
      trees(y).drop(x + 1),
    )
  }

  def visible(x: Int, y: Int, height: Int): Boolean = treesAround(x, y).exists(_.forall(_ < height))

  def scenicScore(x: Int, y: Int, tree: Int): Int = {
    treesAround(x, y).map(line => min(line.size, line.takeWhile(_ < tree).size + 1)).product
  }

  override def part1(): Any = {
    trees.zipWithIndex.map((row: Seq[Int], y: Int) => row.zipWithIndex.count((height: Int, x: Int) => visible(x, y, height))).sum
  }

  override def part2(): Any = {
    trees.zipWithIndex.map((row: Seq[Int], y: Int) => row.zipWithIndex.map((height: Int, x: Int) => scenicScore(x, y, height)).max).max
  }
}
